**Комментарий ревьюера**

Привет, Сергей! Меня зовут Аня Титова, буду проверять твой проект по предобработке. Давай общаться на «ты». Если это неудобно, дай знать, перейдем на «вы».

У тебя уже проделана большая работа над проектом, но давай сделаем его еще лучше. Дальше ты найдешь мои комментарии — пожалуйста, не перемещай и не удаляй их. Буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b> Так выделены небольшие замечания. Одно-два таких замечания в проекте допустимы, но если их будет больше - тебе будет необходимо внести исправления. </div>

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

---

<div class="alert alert-success">
<b>Комментарий ревьюера:</b> Некоторые блоки кода не работают. Посмотри, пожалуйста, что пошло не так. Перед отправкой проекта стоит проверять работоспособность кода - это можно сделать, нажав на панели Jupiter Hub Kernel и Restart & Run All.
    
Еще просмотри, пожалуйста по оформлению: есть дубли заголовков, пустые ячейки и закомментированный код — их стоит убрать из проекта, так будет аккуратнее. А длинные строки кода стоит разбить переносами с помощью \ или скобок, чтобы не было горизонстального скролла — это чтобы было удобнее читать твою работу.
    
<b>Комментарий ревьюера v2:</b> Теперь все ячейки исполняются, а закомментированного кода больше нет. Еще остались дубли заголовков и пустые строки, рекомендую их все же убрать. Пустую ячейку можно удалить кнопкой с ножницами или двойным нажатием на клавишу D. Пустые строки в ячейках с кодом тоже лучше не оставлять, просто чтобы опрятно смотрелось.
 </div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>
 # Не судите сильно строго, это мой первый опыт в подобных проектах и написании кодов, пришлось тратить очень много времени на поиск информации в сети, если есть замечание был бы рад подробно разобрать с тобой. Да Аня, мне тоже было бы комфортно общаться на ты!)



<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Да, постараюсь прокомментировать по шагам. Понимаю, что первую работу может быть сложно выполнять, но уверена, что ты освоишься :) По коду — очень рекомендую шпаргалки, которые даются в тренажере после каждого урока, его должно хватать как раз для выполнения проекта. По оформлению — запроси у своего куратора методичку по проектам, в ней подробный гайд. А гуглить, как удобнее что-то посчитать или красиво вывести — отличная практика, пригодится.
 </div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Заголовок и цель работы есть. Это хоть и очевидные, но важные детали — без них непонятно, что за исследование перед нами. Цель вполне уместно переписывать своими словами, выделяя только самую суть.
 </div>

<div class="alert alert-info"> <b>АНЯ, спасибо за конструктивные замечания и комментарии, постарался все учесть, благодаря им еще раз повторил теорию и даже сделал мини методичку из теории, приведенной в тренажере.) 
    P.S. ...надеюсь я все сделал верно и больше небуду переделывать, но это вопрос только ко мне...)
    заметил один момент, я выполняю определенную задачу, пользуясь теоритическими материалами или нетом и почти сразу забываю, видимо это kun-fu нужно много практиковать...</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> Посмотрим, нужно ли будет что-то доделывать :) Новые комментарии отмечу v3. Перекрашу или зачеркну те, что уже проработаны.
    
Практика как раз позволяет лучше запомнить методы. Думаю, что уроки в тренажере и несколько самостоятельных проектов тебе здорово в этом помогут.
 </div>

<div class="alert alert-success">
<b>Комментарий ревьюера v4:</b> Дальше новые комментарии с отметкой v4. 
 </div>

## Шаг 1. Откройте файл с данными и изучите общую информацию

<div class="alert alert-success">
<b>Комментарий ревьюера v4:</b> Ячейки в твоей тетрадке начинаются не с 1. Рекомендую перед отправкой проекта перезапускать ядро и прогонять все ячейки. Так работа будет опрятнее и проверим работоспособность кода. На панели нажимаем Kernel - Restart & Run All.
 </div>

In [1]:
import pandas as pd
from collections import Counter
from nltk.stem import SnowballStemmer
data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')
display(data.dtypes)



children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Не используй print() на последней строке ячейки, Jupyter Notebook выведет результат и без него. Если в ячейке нужно вывести несколько строк, то для них используем:

* display() для датафреймов и Series
* print() для всего остального
    
А последнюю строку можно не оборачивать в эти функции.    


<b>Комментарий ревьюера v3:</b> Сделано!</div>

**Датасет содержит информацию с данными банка или кредитной организации, содержит все виды файлов
  

Выведим часть датасета, чтоб посмотреть на него.


In [2]:
data.head(10)


children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Проверим в каких столбца отсутствуют данные


In [3]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Отсутствуют значения только в столбцах 'days_employed' и 'total_income'.
Столбцам 'days_employed' и 'total_income' стоит присвоить целочисленные значения (текущая точность нам не нужна)



Проверим, это одни и те же строки, где отсутствуют данные по обоим столбцам ('days_employed' и 'total_income') или нет.

In [4]:
data[(data['total_income'].isnull() == True) & (data['days_employed'].isnull() == True)].info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
children            2174 non-null int64
days_employed       0 non-null float64
dob_years           2174 non-null int64
education           2174 non-null object
education_id        2174 non-null int64
family_status       2174 non-null object
family_status_id    2174 non-null int64
gender              2174 non-null object
income_type         2174 non-null object
debt                2174 non-null int64
total_income        0 non-null float64
purpose             2174 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB



В строках, где отсутствуют данные в столбце 'days_employed', отсутствуют данные и по 'total_income'.

<div class="alert alert-success">
<b>Комментарий ревьюера v4:</b> Верно 
 </div>

Проверим, из разных ли это профессий люди (если из одной, то можно будет заполнить NaN средним по этой группе).


In [5]:
print(data.isnull().mean())


children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64


<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Отличный способ оценить сразу долю пропусков в каждом столбце.</div>

In [6]:
    (print(data[(data['total_income'].isnull() == True) & (data['days_employed']
        .isnull() == True)]['income_type'].value_counts()))

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64


Значения отсутствуют в столбцах total_income и days_employed.
Проверим это люди из разных сфер (прфессий или из одной), ели из одно, то можно весь столбец заменить на срдене арефметическое. 


Люди из разных групп, значит взять среднее по одной группе профессий ('income_type') уже нельзя, будем заполнять NaN по среднему из каждой группы.

<div class="alert alert-success">
<b>Комментарий ревьюера v4:</b> 👍 
 </div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Вот здесь слишком длинная строка, стоит перенести. Например, так:
    
    print(data[(data['total_income'].isnull() == True) & (data['days_employed'] \
        .isnull() == True)]['income_type'].value_counts())
Или так:
    
    (print(data[(data['total_income'].isnull() == True) & (data['days_employed']
        .isnull() == True)]['income_type'].value_counts()))
    
Убираем print(), потому что без него запись проще.
И убираем пустую строку в конце ячейки.

<b>Комментарий ревьюера v3:</b> Отлично, сделано!</div>

In [7]:
from collections import Counter
data.describe( include='all')


children  days_employed     dob_years education  education_id  \
count   21525.000000   19351.000000  21525.000000     21525  21525.000000   
unique           NaN            NaN           NaN        15           NaN   
top              NaN            NaN           NaN   среднее           NaN   
freq             NaN            NaN           NaN     13750           NaN   
mean        0.538908   63046.497661     43.293380       NaN      0.817236   
std         1.381587  140827.311974     12.574584       NaN      0.548138   
min        -1.000000  -18388.949901      0.000000       NaN      0.000000   
25%         0.000000   -2747.423625     33.000000       NaN      1.000000   
50%         0.000000   -1203.369529     42.000000       NaN      1.000000   
75%         1.000000    -291.095954     53.000000       NaN      1.000000   
max        20.000000  401755.400475     75.000000       NaN      4.000000   

          family_status  family_status_id gender income_type          debt  \
count             21525      21525.000000  21525       21525  21525.000000   
unique                5               NaN      3           8           NaN   
top     женат / замужем               NaN      F   сотрудник           NaN   
freq              12380               NaN  14236       11119           NaN   
mean                NaN          0.972544    NaN         NaN      0.080883   
std                 NaN          1.420324    NaN         NaN      0.272661   
min                 NaN          0.000000    NaN         NaN      0.000000   
25%                 NaN          0.000000    NaN         NaN      0.000000   
50%                 NaN          0.000000    NaN         NaN      0.000000   
75%                 NaN          1.000000    NaN         NaN      0.000000   
max                 NaN          4.000000    NaN         NaN      1.000000   

        total_income  purpose  
count   1.935100e+04    21525  
unique           NaN       38  
top              NaN  свадьба  
freq             NaN      797  
mean    1.674223e+05      NaN  
std     1.029716e+05      NaN  
min     2.066726e+04      NaN  
25%     1.030532e+05      NaN  
50%     1.450179e+05      NaN  
75%     2.034351e+05      NaN  
max     2.265604e+06      NaN

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Если передать в describe() параметр include='all', то можно получить информацию еще по категориальным переменным.</div>



1. Минимальное значение по 'children' - -1. Вероятно, это просто отсутствующая информация (хотя это ведь кредиты), скорей всего неверный перенос
2. Максимальное значение по 'children' - 20. Выглядит не очень реалистично,  проверим сколько таких строк имеется в датафрейме.    Вполне возможно, что это ошибка.
2. Отрицательные значения по количеству отработанных дней 'days_employed' - вполне возможно, что если возьмем значения по модулю, то данные станут пригодными для работы.
3. Среднее значение по количеству отработанных дней 'days_employed'  63046 дней (173 года). Кто-то нам нарушает корректность картины.
4. Возраст по некоторым строкам 'dob_years' равен 0 - предполагаю, что просто не указан (хотя, опять же, это ведь кредиты!).



Нужно разобраться со всеми этими некорректными дынными


Выведем некоторые данные по столбцу 'days_employed' чтоб понять в чем ошибка по пункту 3 и 4 в гепотезе выше:

In [8]:

print("Количество строк с 'days_employed' > 0:", data[data['days_employed'] > 0].shape[0])
print("Количество строк с 'days_employed' < 0:", data[data['days_employed'] < 0].shape[0])
data_pens = data[(data['days_employed'] > 0) & (data['income_type'] == 'пенсионер')]
print("Количество пенсионеров с 'days_employed' > 0:", data_pens.shape[0])
print("Количество пенсионеров с 'days_employed' > 0 и кол-вом отработанных лет > 20:", data_pens[data_pens['days_employed'] > 33000].shape[0])
print("Среднее количество отработанных дней среди пенсионеров:", data_pens['days_employed'].mean())

Количество строк с 'days_employed' > 0: 3445
Количество строк с 'days_employed' < 0: 15906
Количество пенсионеров с 'days_employed' > 0: 3443
Количество пенсионеров с 'days_employed' > 0 и кол-вом отработанных лет > 20: 3443
Среднее количество отработанных дней среди пенсионеров: 365003.4912448612



Выведем некоторые данные по столбцам 'dob_years' и 'children', чтоб разобраться с с пунктом 1 и 2

In [9]:

print("Количество людей с нулевым возрастом:", data[data['dob_years'] == 0].count()[0])
print("Количество людей с -1 ребенком:", data[data['children'] == -1].count()[0])
print("Количество строк с 20 детьми:", data[data['children'] == 20].count()[0])
print("Количество уникальных людей с 20 детьми:", len(data[data['children'] == 20]['total_income'].unique()))

Количество людей с нулевым возрастом: 101
Количество людей с -1 ребенком: 47
Количество строк с 20 детьми: 76
Количество уникальных людей с 20 детьми: 68


1)  более 70% значений из 'days_employed' отрицательные и не относятся к пенсионерам - просто возьмем весь столбец по модулю, чтоб убрать;
2) остальные практически все значения > 0 (кроме двух) - это пенсионеры, причем по всем из них стаж работы >  90 лет, что,  выглядит не правдаподобным;
3) нулевой возраст для некоторых клиентов - предполагаю,что данные просто были не внесены в базу, необходимо исправлять;
4) "-1 ребенок" - к исправлению;
5) "20 детей" - это не уникальный случай, а просто ошибка (может быть, корректное значение - "2"). Тоже к исправлению.

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> <s>Прямо в этой ячейке ты сходу и без объясненй заполняешь пропуски в столбцах по стажу и доходу. Стоит сначала исследовать и описать пропущенные значения, и только затем принять решение о том, как их заполнять.
    
Кстати, кому-то из клиентов 0 лет. А не умножаешь ли ты нигде средний стаж на этот 0? Ошибки в столбцах, которые ты используешь для расчетов, стоит устранить до выполнения этих расчетов.</s>


<b>Комментарий ревьюера v4:</b> Отлично, обработку некорректных значений перенесли. В целом на этом шаге нам достаточно посмотреть на общую информацию о данных и решить, нужно предобработка или нет. А все исследование уже делаем на шаге предобработки.</div>
 
 
<div class="alert alert-info">  </b> Тогда я немного переделаю этот раздел...</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера v2:</b> Ты выдвигаешь гипотезы по данным и делаешь проверки — это отлично. Но нужно подвести по ним краткий итог и описать, как поступим с данными на основе этих выводов. 

Пока интерпретации проверок выше нет, а есть сразу заполнение пропусков. До этого этапа нужно пропуски сначала исследовать:

1. Определить тип пропусков (полностью случайный, случайный, неслучайный)
2. Поразмышлять о возможных причинах появления пропусков
3. На основании исследования описать, как будем их заполнять.

<b>Комментарий ревьюера v3:</b> Есть описание пропусков, которые обнаружили. Но ты обходишь стороной их тип. Попробуй посмотреть статью из доп материалов тренажера, возможно она сделает тему понятнее:

https://loginom.ru/blog/missing</div>

1. Попарные NaN 'days_employed' и 'total_inome' заполним значением по данной группе 'type' в соотношении с возрастом.
2. Обрабатываем отрицательные значения по количеству отработанных дней 'days_employed'
3. Столбец 'education' имеет не однотипные значения (напр., 'Среднее'-'среднее'-'СРЕДНЕЕ') - необходимо будет привести к общему виду.
4. Исправляем минимальное значение по 'children' (-1) на 0, а также приведем в соответствие кол-во детей 20...
5. Обрабатываем нулевой возраст по некоторым строкам 'dob_years'.
6. 'days_employed' по пенсионерам > 90 лет - надо приводить к нормальным значениям.

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Кстати, лучше само заполнение пропусков выполнить в разделе «Обработка пропусков», так будет логичнее.</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Мы видим, что есть отрицательные и еще слишком большие значения стажа. Лучше сначала обработать эти ошибки, а только потом рассчитывать характерные значения для заполнения пропусков.

<b>Комментарий ревьюера v3:</b> <s>Замечание актуально.</s>

<b>Комментарий ревьюера v4:</b> Скорректировали план предобработки.</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> Шаги предобраотки намечены.
 </div>

## Шаг 2. Предобработка данных

In [10]:

grouped_data = data.groupby('income_type').agg({'days_employed': 'median', 'total_income': 'median'})
def mean_fill(row):
    if pd.notnull(row['total_income']):
        return row['total_income']
    else:
        return grouped_data[row['income_type']]['total_income']
print(grouped_data)

                 days_employed   total_income
income_type                                  
безработный      366413.652744  131339.751676
в декрете         -3296.759962   53829.130729
госслужащий       -2689.368353  150447.935283
компаньон         -1547.382223  172357.950966
пенсионер        365213.306266  118514.486412
предприниматель    -520.848083  499163.144947
сотрудник         -1574.202821  142594.396847
студент            -578.751554   98201.625314


<div class="alert alert-success">
<b>Комментарий ревьюера v4:</b> Здесь видим, что тип дохода здорово описывает медианный доход. А со стажем не все так ясно, скорее здесь будет уместнее другая характеристика. 
 </div>

### Обработка пропусков

Берем значения по 'days_employed' и 'total_income' по модулю.
Добавляем столбец 'ratio_days_employed' для вычисления коэффициента полноты количества отработанных дней (учитывая то, что официально можно трудиться с 16 лет), чтобы впоследствии заполнить NaN по 'days_employed' с учетом среднего этого показателя по группе.

<div class="alert alert-warning">
<b>Комментарий ревьюера v4:</b> А разве в столбце с доходом были отрицательные значения? 
 </div>

In [11]:
data[['total_income', 'days_employed']] = data[['total_income', 'days_employed']].abs()
data['ratio_days_employed'] = data[data['days_employed'].notnull()]['days_employed']/((data['dob_years']-16)*365)

data.head()


children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  ratio_days_employed  
0               покупка жилья             0.889112  
1     приобретение автомобиля             0.551343  
2               покупка жилья             0.906273  
3  дополнительное образование             0.706292  
4             сыграть свадьбу            25.195563

Заполним пропуски NaN в датафрейме data:
-'days_employed': NaN заменяем на произведение среднего коэффициента по группе 'income_type', умноженного на возраст в днях;
-'total_income': NaN заменяем на среднее значение по зарплате по группе 'income_type'.
удалим вспомогательный столбец


In [12]:
data['days_employed'] = data.groupby('income_type')['days_employed'].transform(lambda x: x.fillna(x.mean()*data['dob_years']*365))
data['total_income'] = data.groupby('income_type')['total_income'].transform(lambda x: x.fillna(x.mean()))
data = data.drop('ratio_days_employed', axis=1)
data.head()


children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<div class="alert alert-success">
<b>Комментарий ревьюера v4:</b> Отлично. Аномальные значения здесь допустимо оставить, потому что они встречаются только у двух категорий клиентов, а мы заполняем как раз средним по категории.
 </div>

Кол-во детей с отрицательным значением заменим на среднюю методом abs(), "ошибку" ввиде 20 детей, щаменим на 2

In [13]:
data['children'] = data['children'].abs()

data.loc[data['children']>10, 'children']=2

data.head()


children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> Ок, отрицательные значения и проблемы с количеством детей устранили.
 </div>

Проверяем данные

In [14]:
data.isnull().mean()

children            0.0
days_employed       0.0
dob_years           0.0
education           0.0
education_id        0.0
family_status       0.0
family_status_id    0.0
gender              0.0
income_type         0.0
debt                0.0
total_income        0.0
purpose             0.0
dtype: float64

после проверки пропусков не выявлено

In [15]:
print(data[(data['total_income'].isnull() == True) & (data['days_employed'] \
    .isnull() == True)]['income_type'].value_counts())

Series([], Name: income_type, dtype: int64)


Проверили данные двумя методами, пропусков не осталось

Cмотрим количество нулей по возрасту (столбец 'dob_years') и количество несовершеннолетних


In [16]:
# смотрим количество нулей по возрасту (столбец 'dob_years') и количество несовершеннолетних
print("Количество строк с 'dob_years' 0 -", data[data['dob_years'] == 0].count()[0])
print("Количество строк с 'dob_years' < 19 -", data[data['dob_years'] < 19].count()[0])

Количество строк с 'dob_years' 0 - 101
Количество строк с 'dob_years' < 19 - 101


Количество совпадает? это одни и те же люди. Заполним значения средним по каждой группе

In [17]:
data['dob_years'] = data.groupby('income_type')['dob_years'].transform(lambda x: x.replace(0, int(x.mean())))



In [18]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   2.152500e+04  21525.000000  21525.000000   
mean       0.479721   1.522834e+08     43.495238      0.817236   
std        0.755528   1.078074e+09     12.230322      0.548138   
min        0.000000   0.000000e+00     19.000000      0.000000   
25%        0.000000   1.024652e+03     34.000000      1.000000   
50%        0.000000   2.605748e+03     43.000000      1.000000   
75%        1.000000   3.336411e+05     53.000000      1.000000   
max        5.000000   9.725518e+09     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.673959e+05  
std            1.420324      0.272661  9.790695e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.519313e+05  
75%            1.000000      0.000000  2.024175e+05  
max            4.000000      1.000000  2.265604e+06

Обработали пропуски, убрали некооректные данные по кол-ву детей, также откоррктировали отрицательный стаж, у нас остался неккоректный стаж по пенсионерам, причины этой ошибки выяснить тяжело, возможно при переносе данных или некорректное заполение. Если у нас нет возможности уточнить эти данные, можно предположить, что теоретически стаж пенсионеров должен быть больше,чем у остальных категорий(за исключением военных и структу МВД, но они редко выходят на пенсию с наступлением пенсионного возраста, а продолжают работать), поэто если нам это будет нужно этот столбец можно категоризировать например стаж: более 50 лет и все что выше попадет в эту категорию.
Думамое можно сделать вывод, что основные все явно некорректные значения откорректированы.

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> <s>В этом разделе есть только код, и нет пояснений, что он делает. Прокомментируй, пожалуйста, каждый шаг.</s>

<b>Комментарий ревьюера v3:</b> Отлично, комментарии есть!</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Помимо явных пропусков, которые мы нашли с помощью isnull(), в данных могут быть скрытые, замененные значениями по умолчанию. Попробуй их найти и устранить. Для категориальных переменных тебе пригодится метод value_counts().

<b>Комментарий ревьюера v3:</b> Часть значений обрабтали, но ты ничего не сделал с 0 лет у клиентов и аномально большим стажем у пенсионеров. С возрастом нужно разобраться, потому что ты используешь столбец для заполнения пропусков. Стаж на твое усмотрение, но если не будешь обрабатывать, то дай пояснение, почему.

<b>Комментарий ревьюера v4:</b> Отлично, всё сделано! Описать, почему не корректируешь некоторые значения — тоже возможный вариант. Но тут дам подсказку. Обычно в базе данных хранят дату в unix time, а нам выгрузили в количестве дней. При переводе из одного формата в другой можно ошибиться с коэффициентами.</div>

<div class="alert alert-info"> <b>Вроде учел все рекомендации</b>

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> Улучшения значительные, но пока еще есть, чем заняться :)
 </div>

### Замена типа данных

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Дубль заголовков и пустая ячейка под ними — стоит поправить оформление.</div>

In [19]:
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')
data.info()
data.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   СРЕДНЕЕ             1   
8         2           6929         35    ВЫСШЕЕ             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Типы данных привели. Кстати, лучше описывать, что мы собираемся делать перед ячейкой кода, чем постфактум после. После уместо сделать кратикй вывод по проделанной работе. Прям по каждой ячейке не нужно, но в конце каждого шага — вполне.</div>

месячную зп и стаж перевели в целые числа, таблица стала более читаемая, приведем столбец education к одному регистру, а то он "бросается" в глаза.

In [20]:
data['education'] = data['education'].str.lower()
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Постарайся не вызывать весь датафрейм целиком — так можно повесить Jupyter Hub. Лучше используй для вывода первых и последних строк методы head() и tail().

<b>Комментарий ревьюера v3:</b> 👍</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Мы не просто привели образование к единообразной записи, но и убрали дубликаты из-за разного регистра.</div>

<div class="alert alert-info"> <b>Рекомендацию ниже тоже учел, подскажите есть ли какое-то стандартное сочетание клавиш для выделения синим</b>

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> Шортката, тем более стандартного, нет. Можно справиться с помощью сниппетов, но я сама не углублялась в тему :)
 </div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Отрицательные значения стажа обработали, но лучше это сделать до заполнения пропусков. Иначе мы рассчитали характерыне значения уже с этим ошибками. Отдельно посмотри на значения вроде 340266 — это ведь больше 900 лет. Скорее всего, ошибка системная, и ее можно исследовать и устранить. Сделать то и другое лучше на этапе обработки пропусков. Почему — сейчас эти значения содержат ошибки, поэтому они для нас все равно что пропуск информации. Но исходные значения можно восстановить. 

Эта часть не обязательная, потому что столбец со стажем может нам не понадобиться в исследовании. Но хорошая практика — очистить от ошибок весь датафрейм.

<b>Комментарий ревьюера v3:</b> <s>Эта часть актуальна. Ячейка, в которой ты заполняешь пропуски, у тебя расположена в самом начале проверок, о устранения каких-либо ошибок в данных.</s>

<b>Комментарий ревьюера v4:</b> Здесь всё готово.</div>

### Обработка дубликатов

In [21]:
data.duplicated().sum()

71

Посмотрим на дубликаты

In [22]:
data[data.duplicated(keep=False)].sort_values(by=['income_type', 'days_employed'])

children  days_employed  dob_years education  education_id  \
680           1       37228871         30    высшее             0   
18349         1       37228871         30    высшее             0   
9013          2       44674645         36    высшее             0   
14432         2       44674645         36    высшее             0   
8490          1       23891898         31   среднее             1   
...         ...            ...        ...       ...           ...   
12373         0       49251988         58   среднее             1   
15091         0       49251988         58   среднее             1   
20662         0       49251988         58   среднее             1   
554           0       50950332         60   среднее             1   
15188         0       50950332         60   среднее             1   

          family_status  family_status_id gender  income_type  debt  \
680     женат / замужем                 0      F  госслужащий     0   
18349   женат / замужем                 0      F  госслужащий     0   
9013    женат / замужем                 0      F  госслужащий     0   
14432   женат / замужем                 0      F  госслужащий     0   
8490    женат / замужем                 0      F    компаньон     0   
...                 ...               ...    ...          ...   ...   
12373   женат / замужем                 0      M    сотрудник     0   
15091  гражданский брак                 1      M    сотрудник     0   
20662  гражданский брак                 1      M    сотрудник     0   
554     женат / замужем                 0      M    сотрудник     0   
15188   женат / замужем                 0      M    сотрудник     0   

       total_income                  purpose  
680          170898  покупка жилья для семьи  
18349        170898  покупка жилья для семьи  
9013         170898    получение образования  
14432        170898    получение образования  
8490         202417            покупка жилья  
...             ...                      ...  
12373        161380     покупка недвижимости  
15091        161380    на проведение свадьбы  
20662        161380    на проведение свадьбы  
554          161380     покупка недвижимости  
15188        161380     покупка недвижимости  

[137 rows x 12 columns]

Похоже на бональное задвоение, удалим их.

In [23]:
data = data.drop_duplicates()
print(data.duplicated().sum())

0


Удалили из масива данных дубликаты

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Отлично, дубликаты обработали.</div>

### Лемматизация

В стодбце purpose есть список на первый взгляд похожих целей, выведим список уникальных целей

In [24]:
unique_purposes = data['purpose'].value_counts().index.tolist()
data['purpose'].value_counts().to_frame()

purpose
свадьба                                     791
на проведение свадьбы                       768
сыграть свадьбу                             765
операции с недвижимостью                    675
покупка коммерческой недвижимости           661
операции с жильем                           652
покупка жилья для сдачи                     651
операции с коммерческой недвижимостью       650
покупка жилья                               646
жилье                                       646
покупка жилья для семьи                     638
строительство собственной недвижимости      635
недвижимость                                633
операции со своей недвижимостью             627
строительство жилой недвижимости            624
покупка недвижимости                        621
покупка своего жилья                        620
строительство недвижимости                  619
ремонт жилью                                607
покупка жилой недвижимости                  606
на покупку своего автомобиля                505
заняться высшим образованием                496
автомобиль                                  494
сделка с подержанным автомобилем            486
свой автомобиль                             478
на покупку подержанного автомобиля          478
автомобили                                  478
на покупку автомобиля                       471
приобретение автомобиля                     461
дополнительное образование                  460
сделка с автомобилем                        455
высшее образование                          452
образование                                 447
получение дополнительного образования       446
получение образования                       442
профильное образование                      436
получение высшего образования               426
заняться образованием                       408

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Метод to_frame() позволил бы красиво отобразить эти данные, но нужно убрать print().

<b>Комментарий ревьюера v3:</b> ✨</div>

видно, что есть ряд одинаковых целей по разному интерпретированных, применим метод лемматизации, чтобы выделить уникальные леммы

In [25]:
from collections import Counter
from nltk.stem import SnowballStemmer 
russian_stemmer = SnowballStemmer('russian')

from pymystem3 import Mystem
m = Mystem()


In [26]:

def do_lemma(row):
    lemma = m.lemmatize(row)
    return lemma


Cоздадим дополнительный столбец с лемматизированными целями


In [27]:
data['purpose_lemma'] = data['purpose'].apply(do_lemma)

создаем строку из уникальных целей



In [28]:
all_lemmas_list = data['purpose_lemma'].values
flat_lemmas_list = []
for sublist in all_lemmas_list:
    for item in sublist:
        flat_lemmas_list.append(item)

In [29]:
print(all_lemmas_list)
from collections import Counter
#посчитаем самые популярные слова, чтобы обобщить цели кредитов, это: недвижимость, образование, автомобиль, свадьба
print('Посчитаем самые популярные леммы в графе purpose:')
print(Counter(flat_lemmas_list))

[list(['покупка', ' ', 'жилье', '\n'])
 list(['приобретение', ' ', 'автомобиль', '\n'])
 list(['покупка', ' ', 'жилье', '\n']) ... list(['недвижимость', '\n'])
 list(['на', ' ', 'покупка', ' ', 'свой', ' ', 'автомобиль', '\n'])
 list(['на', ' ', 'покупка', ' ', 'автомобиль', '\n'])]
Посчитаем самые популярные леммы в графе purpose:
Counter({' ': 33570, '\n': 21454, 'недвижимость': 6351, 'покупка': 5897, 'жилье': 4460, 'автомобиль': 4306, 'образование': 4013, 'с': 2918, 'операция': 2604, 'свадьба': 2324, 'свой': 2230, 'на': 2222, 'строительство': 1878, 'высокий': 1374, 'получение': 1314, 'коммерческий': 1311, 'для': 1289, 'жилой': 1230, 'сделка': 941, 'дополнительный': 906, 'заниматься': 904, 'проведение': 768, 'сыграть': 765, 'сдача': 651, 'семья': 638, 'собственный': 635, 'со': 627, 'ремонт': 607, 'подержанный': 486, 'подержать': 478, 'приобретение': 461, 'профильный': 436})


выводим функцию для создания словаря

In [30]:
def dictionary(world, data):
    for row in data:
        if world == row:
            return row

print('Cоздадим словарь с категориями в графе purpose, вдальнейшем его можно будет пополнять:')
#создадим словарь с категориями, вдальнейшем его можно будет пополнять

dict = []
#вручную добавим выбранные мною слова
dict.append(dictionary('жилье', flat_lemmas_list))
dict.append(dictionary('недвижимость', flat_lemmas_list))
dict.append(dictionary('образование', flat_lemmas_list))
dict.append(dictionary('автомобиль', flat_lemmas_list))
dict.append(dictionary('свадьба', flat_lemmas_list))
print(dict)

Cоздадим словарь с категориями в графе purpose, вдальнейшем его можно будет пополнять:
['жилье', 'недвижимость', 'образование', 'автомобиль', 'свадьба']


Создаем столбец с категориями и функцию которая выдает категории

In [31]:
def category(data):
    if dict[0] in data:
        return dict[1]
    elif dict[1] in data:
        return dict[1]
    elif dict[2] in data:
        return dict[2]
    elif dict[3] in data:
        return dict[3]
    else:
        return dict[4]
    

data['category_purpose'] = data['purpose_lemma'].apply(category)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose                              purpose_lemma  \
0               покупка жилья                    [покупка,  , жилье, \n]   
1     приобретение автомобиля          [приобретение,  , автомобиль, \n]   
2               покупка жилья                    [покупка,  , жилье, \n]   
3  дополнительное образование       [дополнительный,  , образование, \n]   
4             сыграть свадьбу                  [сыграть,  , свадьба, \n]   
5               покупка жилья                    [покупка,  , жилье, \n]   
6           операции с жильем             [операция,  , с,  , жилье, \n]   
7                 образование                          [образование, \n]   
8       на проведение свадьбы        [на,  , проведение,  , свадьба, \n]   
9     покупка жилья для семьи  [покупка,  , жилье,  , для,  , семья, \n]   

  category_purpose  
0     недвижимость  
1       автомобиль  
2     недвижимость  
3      образование  
4          свадьба  
5     недвижимость  
6     недвижимость  
7      образование  
8          свадьба  
9     недвижимость

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Не используй циклы для построчной обработки таблицы — это медленно. Лучше напиши функцию для этой операции и передай её в метод apply().
        
        

<b>Комментарий ревьюера v3:</b> <s>Актуально.</s>

<b>Комментарий ревьюера v4:</b> Сделано!</div>

<div class="alert alert-info"> <b>исправил</b> 

In [32]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose                              purpose_lemma  \
0               покупка жилья                    [покупка,  , жилье, \n]   
1     приобретение автомобиля          [приобретение,  , автомобиль, \n]   
2               покупка жилья                    [покупка,  , жилье, \n]   
3  дополнительное образование       [дополнительный,  , образование, \n]   
4             сыграть свадьбу                  [сыграть,  , свадьба, \n]   
5               покупка жилья                    [покупка,  , жилье, \n]   
6           операции с жильем             [операция,  , с,  , жилье, \n]   
7                 образование                          [образование, \n]   
8       на проведение свадьбы        [на,  , проведение,  , свадьба, \n]   
9     покупка жилья для семьи  [покупка,  , жилье,  , для,  , семья, \n]   

  category_purpose  
0     недвижимость  
1       автомобиль  
2     недвижимость  
3      образование  
4          свадьба  
5     недвижимость  
6     недвижимость  
7      образование  
8          свадьба  
9     недвижимость

Мы получили список уникальных лемм, недвижимость и жилье это одно и тоже.

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Лемматизацию выполнили! Но раз в конце ты оставляешь только 4 леммы, можно задать их список вручную.</div>

<div class="alert alert-info"> <b>Исправил</b> 

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> <s>А судя по количеству None в столбце final_purposes, функция и не сработала. </s>


<b>Комментарий ревьюера v4:</b> Теперь порядок. Но стоит проверять, как отработала твоя функция. </div>

### Категоризация данных

In [33]:
debt_from_children = pd.DataFrame()
debt_from_children['count_children'] = data.groupby('children')['debt'].count()
debt_from_children['sum_children'] = data.groupby('children')['debt'].sum()
debt_from_children['result_children'] = debt_from_children['sum_children'] / debt_from_children['count_children'] 
debt_from_children


count_children  sum_children  result_children
children                                               
0                  14091          1063         0.075438
1                   4855           445         0.091658
2                   2128           202         0.094925
3                    330            27         0.081818
4                     41             4         0.097561
5                      9             0         0.000000

Вывод:
Так мы изучаем влияние возврата кредита в срок и наличие детей, видим отрицательное значение и значение 20, считаю, что это опечатка при переносе данных. Категоризация была применена вместе с лемматизацией выше в одной функции. Категоризовать количество детей нет необходимости, категориями будет как раз их количество, так как всего 6 вариантов.

</div>  Я не смог заменить 20 детей на медиану, использовал data['children'] = data['children'].replace(20, children_median), но выдает ошибку, не могу понять почему, тоже самое с -1, это опечатка, как бы вывод могу сделать итак, но хочется заменить эти данные </b>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Нужно сначала задать переменную children_median. А еще эти ошибки стоило обнаружить на этапе обработки пропусков :)</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Кстати, чтобы твои комментарии красиво отображались, помещай их до закрывающего тега:
    
    <div> текст пишем между двух этих тегов </div>
    
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Найти и опиши, какие словари уже есть в данных. Это распространенный тип данных, мне нужно убедиться, что ты им владеешь. Скорее всего на шаге 3 тебе понадобятся дополнительные словари, например, по уровню дохода. Опиши и выдели их тоже в этом разделе.
       
</div>

<div class="alert alert-info"> <b>Я не пойму как это сделать..., но попробую</b> 

In [34]:
def salary_cat(total_income):
    if total_income <=50000:
                    return 'бедный'
    elif 50000 <total_income<=120000:
                    return 'средний'
    elif 120000 <total_income<=250000:
                    return 'выше среднего'
    else:
                    return 'богатый'
salary_cat


<function __main__.salary_cat(total_income)>

In [35]:
data['salary_cat'] = data['total_income'].apply(salary_cat)
data.head(10)


children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose                              purpose_lemma  \
0               покупка жилья                    [покупка,  , жилье, \n]   
1     приобретение автомобиля          [приобретение,  , автомобиль, \n]   
2               покупка жилья                    [покупка,  , жилье, \n]   
3  дополнительное образование       [дополнительный,  , образование, \n]   
4             сыграть свадьбу                  [сыграть,  , свадьба, \n]   
5               покупка жилья                    [покупка,  , жилье, \n]   
6           операции с жильем             [операция,  , с,  , жилье, \n]   
7                 образование                          [образование, \n]   
8       на проведение свадьбы        [на,  , проведение,  , свадьба, \n]   
9     покупка жилья для семьи  [покупка,  , жилье,  , для,  , семья, \n]   

  category_purpose     salary_cat  
0     недвижимость        богатый  
1       автомобиль        средний  
2     недвижимость  выше среднего  
3      образование        богатый  
4          свадьба  выше среднего  
5     недвижимость        богатый  
6     недвижимость  выше среднего  
7      образование  выше среднего  
8          свадьба        средний  
9     недвижимость  выше среднего

Категория бедный не выпадает если выводить хоть 10 первых, хоть 10 последних строк, выведим её отдельно, чтоб потом сделать вывод, на что берут кредит люди в этой категории.

In [36]:
data = data.set_index(['salary_cat'])
data.loc[data.index.isin(['бедный'])]

children  days_employed  dob_years education  education_id  \
salary_cat                                                               
бедный             0           1181         54   среднее             1   
бедный             0         338113         62   среднее             1   
бедный             0             72         42   среднее             1   
бедный             0         382189         60   среднее             1   
бедный             1           3597         39   среднее             1   
...              ...            ...        ...       ...           ...   
бедный             0         358631         71   среднее             1   
бедный             0           1446         45   среднее             1   
бедный             0         338060         59   среднее             1   
бедный             0         365213         52   среднее             1   
бедный             2           1867         32   среднее             1   

                    family_status  family_status_id gender income_type  debt  \
salary_cat                                                                     
бедный            женат / замужем                 0      F   сотрудник     0   
бедный            женат / замужем                 0      F   пенсионер     0   
бедный           гражданский брак                 1      F   компаньон     0   
бедный      Не женат / не замужем                 4      F   пенсионер     0   
бедный            женат / замужем                 0      F   компаньон     0   
...                           ...               ...    ...         ...   ...   
бедный            женат / замужем                 0      F   пенсионер     0   
бедный            женат / замужем                 0      F   сотрудник     0   
бедный            женат / замужем                 0      F   пенсионер     0   
бедный            женат / замужем                 0      F   пенсионер     1   
бедный            женат / замужем                 0      F   сотрудник     0   

            total_income                                purpose  \
salary_cat                                                        
бедный             44591                             автомобиль   
бедный             43929                             автомобили   
бедный             49463       сделка с подержанным автомобилем   
бедный             31534      покупка коммерческой недвижимости   
бедный             40891                   покупка своего жилья   
...                  ...                                    ...   
бедный             41411                 профильное образование   
бедный             34472                покупка жилья для сдачи   
бедный             47388                  получение образования   
бедный             48239                        свой автомобиль   
бедный             34767  операции с коммерческой недвижимостью   

                                                purpose_lemma category_purpose  
salary_cat                                                                      
бедный                                       [автомобиль, \n]       автомобиль  
бедный                                       [автомобиль, \n]       автомобиль  
бедный      [сделка,  , с,  , подержанный,  , автомобиль, \n]       автомобиль  
бедный        [покупка,  , коммерческий,  , недвижимость, \n]     недвижимость  
бедный                       [покупка,  , свой,  , жилье, \n]     недвижимость  
...                                                       ...              ...  
бедный                       [профильный,  , образование, \n]      образование  
бедный              [покупка,  , жилье,  , для,  , сдача, \n]     недвижимость  
бедный                        [получение,  , образование, \n]      образование  
бедный                              [свой,  , автомобиль, \n]       автомобиль  
бедный      [операция,  , с,  , коммерческий,  , недвижимо...     недвижимость  

[372 rows x 14 columns]

<div class="alert alert-info"> <b>Вроде получилось, если я правильно понял твой посыл Аня :)</b> 

In [37]:
print('Создали словарь:')
family_dict = data[['family_status_id', 'family_status']]
family_dict = family_dict.drop_duplicates().reset_index(drop=True)
display(family_dict)
print('\n\nСгруппированная таблица. Берем по id, другой столбец удалили:')
a = data.groupby('family_status_id')['debt'].agg(['count', 'sum', 'mean'])
display(a)
# Заменяем
print('\n\nЗаменяем численные значения по ключу словаря:')
a.reset_index().replace({'family_status_id': family_dict.family_status.to_dict()})

Создали словарь:


family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем



Сгруппированная таблица. Берем по id, другой столбец удалили:


count  sum      mean
family_status_id                      
0                 12339  931  0.075452
1                  4151  388  0.093471
2                   959   63  0.065693
3                  1195   85  0.071130
4                  2810  274  0.097509



Заменяем численные значения по ключу словаря:


family_status_id  count  sum      mean
0        женат / замужем  12339  931  0.075452
1       гражданский брак   4151  388  0.093471
2         вдовец / вдова    959   63  0.065693
3              в разводе   1195   85  0.071130
4  Не женат / не замужем   2810  274  0.097509

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> Самое важное здесь — описать. Почему выделили категории именно так? Сейчас есть только код, без комментариев.
    
По существующим словарям в данных подскажу: есть пары столбцов id - расшифровка. Постарайся их найти в наших данных и вывести на экран. 

<b>Комментарий ревьюера v4:</b> Словари нашли :)</div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [38]:
debt_from_children = pd.DataFrame()
debt_from_children['count_children'] = data.groupby('children')['debt'].count()
debt_from_children['sum_children'] = data.groupby('children')['debt'].sum()
debt_from_children['result_children'] = debt_from_children['sum_children'] / debt_from_children['count_children'] 
debt_from_children.sort_values('result_children', ascending = False)

count_children  sum_children  result_children
children                                               
4                     41             4         0.097561
2                   2128           202         0.094925
1                   4855           445         0.091658
3                    330            27         0.081818
0                  14091          1063         0.075438
5                      9             0         0.000000

Вывод: кредитору, не имеющие детей менее склонны к просрочке по кредитам, видимо сказывается финансовая нагрузка. Но также видно,что люди у кого три ребенка реже просрачевают, чем у кого три ребенка, возможно еще влияет возвраст детей, но у нас нет этих данных.


<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Ты используешь правильную метрику — количество должников в группе. Но ее можно рассчитать проще. В столбце debt только 0 и 1, поэтому применяем к нему метод mean() и получаем долю 1, то есть должников.
    
Выводы по клиентам без детей правильные, а вот про клиентов с 3 деьми мы не можем точно утверждать, что у них реже просрочки. Посмотри, насколько малочисленна эта группа. Скорее всего по ней просто мало данных.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> <s>Ошибочные значения в количестве детей нужно убрать перед тем, как выполнять этот расчет.</s>


<b>Комментарий ревьюера v3:</b> Сделано! </div>
<div class="alert alert-info"> <b>Вроде исправил :)</b> 


- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [39]:
debt_from_family_status = pd.DataFrame()
debt_from_family_status['sum_family_status'] = data.groupby('family_status')['debt'].sum()
debt_from_family_status['count_family_status'] = data.groupby('family_status')['debt'].count()
debt_from_family_status['result_family_status'] = debt_from_family_status['sum_family_status'] / debt_from_family_status['count_family_status'] 
debt_from_family_status.sort_values('result_family_status', ascending = False)

sum_family_status  count_family_status  \
family_status                                                   
Не женат / не замужем                274                 2810   
гражданский брак                     388                 4151   
женат / замужем                      931                12339   
в разводе                             85                 1195   
вдовец / вдова                        63                  959   

                       result_family_status  
family_status                                
Не женат / не замужем              0.097509  
гражданский брак                   0.093471  
женат / замужем                    0.075452  
в разводе                          0.071130  
вдовец / вдова                     0.065693

**Вывод: семейный статус влияент на уплату кредита в срок, люди не сосотящие в браке или находящиеся в гражданском браке обладают более худщей платежной дисциплиой, но люди которые побывали в браке, лучше возвращают долги, чем те кто не состоял в отношениях совсем)

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Ок.
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [40]:
def relation(category):
    return data.groupby(category)['debt'].mean().to_frame().sort_values(by='debt')

relation('salary_cat')

debt
salary_cat             
бедный         0.061828
богатый        0.068966
средний        0.081570
выше среднего  0.084435

Вывод:
Люди с доходом более 50 000 руб. чаще просрачивают кредит.

Но если перепроверить данные, построив другую функцию, то получим

In [41]:
print('Разобьем доход на уровни: 0,20-50, 50-100, 100-250, 250-500, 500+')

def income(data):
    if data == 0:
        return '0'
    elif data <= 50000:
        return '20-50'
    elif data <= 100000:
        return '50-100'
    elif data <= 250000:
        return '100-250'
    elif data <= 500000:
        return '250-500'
    else:
        return '500+'
#

data['category_income'] = data['total_income'].apply(income)
data_pivot_income = data.pivot_table(index='category_income', columns='debt', values='total_income', aggfunc='count')
#считаем процент должников
data_pivot_income['ratio %'] = (data_pivot_income[1] / data_pivot_income[0]) * 100
data_pivot_income['ratio %'] = data_pivot_income['ratio %'].astype('int')
#cчитаем сумму клиентов каждой категории
data_pivot_income['sum'] = data_pivot_income[1] + data_pivot_income[0]

data_pivot_income


Разобьем доход на уровни: 0,20-50, 50-100, 100-250, 250-500, 500+


debt                 0     1  ratio %    sum
category_income                             
100-250          12985  1193        9  14178
20-50              349    23        6    372
250-500           2411   180        7   2591
50-100            3760   331        8   4091
500+               208    14        6    222

Вывод: люди с доходом 50 000 руб. все еще самые ответственные заемщики, но также к ним прибавились люди с дохожом от 500 000 руб, видимо сказыается цель кредита и т.к. мы выводили таблицу именно по этой категории, я бы выдвинул две гипотезы:
1. в этой категории много пенсионеров, а это чаще люди более ответсвенные в силу своего возраста и воспитания.
2. это люди, берущие займ на недвижимость и не хотят допускать просрочки, т.к. есть риски, например банк может затребовать погасить сразу всю сумму,  что для них будет фатальным, ну и видимо, не хотят портить историю,т.к. скорей всего это не последний кредит в их жизни.)
3. что касатся людей с доходом 500 000+, с таким уровнем ежемесчный платеж комфортный и позволяет не допускать просрочки, возможно стоит автоплатеж.

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Для того, чтобы убрать такие предупреждения, можно использовать вот такой код:
    
    import warnings
    warnings.filterwarnings("ignore")
    
<div class="alert alert-info"> <b>Спасибо, так выглядит лучше, ножно почитать, что мы сделали, не помню этот прием в теоритечиских занятиях</b>
       

<div class="alert alert-success">
<b>Комментарий ревьюера v3:</b> Предупреждения часто зависят от версии используемой библиотеки. А в тренажере обычно дается такое задание, чтобы код отрабатывал без предупреждений.
 </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера v2:</b> 10 групп — это слишком много, среди них есть слишком похожие. Оптимально будет выделить 3-5 групп. Также нужно описать, как ты их выделяешь. Лучше всего сделать это на этапе категоризации.
    
<b>Комментарий ревьюера v3:</b> Ты ведь выделил три группы на этапе категоризации. Для чего здесь продожаешь использовать 10?

<b>Комментарий ревьюера v4:</b> Ок, уменьшили количество групп. Но на предыдущем шаге мы уже создали столбец salary_cat с категориями по доходу. Его и следует здесь использовать. </div>

- Как разные цели кредита влияют на его возврат в срок?

In [42]:
data_pivot_purpose = data.pivot_table(index='category_purpose', columns='debt', values='purpose', aggfunc='count')
data_pivot_purpose['ratio %'] = (data_pivot_purpose[1] / data_pivot_purpose[0]) * 100
data_pivot_purpose

debt                  0    1    ratio %
category_purpose                       
автомобиль         3903  403  10.325391
недвижимость      10029  782   7.797388
образование        3643  370  10.156464
свадьба            2138  186   8.699719

**Вывод:
Заемщики, берущие кредит для приобретения/проведение операций с жильем, наиболее ответственны и менее склонны нарушать обязательства по выплатам кредита в срок.**

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Используй в качестве категорий леммы, которые ты определил ранее.
    
<b>Комментарий ревьюера v3:</b> Сделано только вот данных не хватает - это видно по столбцам с count. Категории определились не для всех записей.

<b>Комментарий ревьюера v4:</b> Готово.</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Ты отлично владеешь методом groupby(), но иногда нам удобно объединять данные с помощью pivot_table(). Это полезный метод, и он тебе не раз пригодится. Мне нужно убедиться, что ты умеешь пользоваться и им тоже. Продемонстрируй, пожалуйста, на примере одного из расчетов в этом шаге.
    
<b>Комментарий ревьюера v3:</b> Сделано! Вывод не должен поменяться, это просто способ получить ту же таблицу иначе. У pivot_table() немного другие параметры, и иногда собрать данные воедино получается только им.</div>

   
<div class="alert alert-info"> <b>Учел замечнаие по сводной таблице, вывод не поменялся</b> 

## Шаг 4. Общий вывод

Отвечая на поставленный вопрос "Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок", могу дать следующий ответ - семеное положение и количество детей влияет на факт погашения кредита в строк:
-заемщики с официально оформленными отношениями (или которые в прошлом были в официальном в браке) и не имеющие детей - самые ответственные заемщики;
-заемщики, состоящие в неофициальном браке или находящиеся без отношений, при этом имеющие 1 или 2 детей -  менее ответственные заемщики.


ТАКЖЕ МОЖНО ДОБАВИТЬ:

Ответственный заещик	Не имеет детей,	находится/был в официальном браке	Уровень дохода < 50000	Берет кредит на жилье или пенсионер, который обладает более лучшей платежной дисциплиной.
Менее ответственный заещик	Имеет 1-2 детей	Не в узаконенных отношениях/не в отношениях	Уровень дохода > 50000	Берет кредит на образование/свадьбу. Также ответственный плательщик с доходом +500 000

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b> Здорово, что упоминаешь поставленную задачу и даешь рекомендации как по клиентам с наибольшим, так и с наименьшим количеством просрочек. Но мы в предыдущем шаге ответили на вопросы по 4 категориям, стоит в выводе привести все эти результаты.
    
</div>

---

<div class="alert alert-success">
<b>Итоговый комментарий ревьюера v2:</b> Ты справился со множеством задач в этом проекте, которые могут вызывать затруднение, молодец! Пока еще есть, что доделать: разобрала пару примеров с оформлением — постарайся учесть эти замечания по всему проекту. И оставила комментарии к шагам, где нужно что-то поправить. Внеси, пожалуйста, исправления, и присылай работу на дополнительное ревью :)
    
</div>

<div class="alert alert-success">
<b>Итоговый комментарий ревьюера v3</b> Проект стал подробнее и полнее — отлично. Сейчас тоже остались замечания, которые нужно учесть. Соберу в чек-лист. Обязательные:
    
* заполняешь пропуски на этапе, где этого делать не нужно. Возможно, случайно, но стоит разобраться.
* функция категоризации целей кредита не работает
* словари нужно описать, а не просто привести. 

Небольшие замечания, но пока их многовато:

* определяем тип пропусков
* принимаем решение по нулевому возрасту и аномальному стажу
* функция с apply вместо построчной обработки датафрема циклом
* для исследования уровня дохода и просрочек следует использовать 3-5 групп 
    
</div>

<div class="alert alert-info"> <b> Аня, постарался максимально учесть все твои замечания.)</b> 

<div class="alert alert-success">
<b>Итоговый комментарий ревьюера v4</b> Почти все замечания действительно учтены, а предобработку подняли на новый уровень! На этом могу принять проект. Спасибо за работу :)
    
Из рекомендаций на будущее: постарайся разобраться с типами пропусков. В следующих проектах будут встречаться как достаточно чистые датасеты, так и требующие обработки. Понимание, что за пропуск перед тобой, здорово поможет в принятии решения по нему.</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [X]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.